---
title: "Geuvadis"
author: "Armin Rauschenberger"
date: "`r format(Sys.time(),'%d %B %Y')`"
output:
    html_document:
        theme: default
        highlight: textmate
        code_folding: hide
---

```{r knitr,include=FALSE}
knitr::opts_chunk$set(eval=FALSE)
```

********************************************************************************
rm(list=ls());gc()

# __1. Initialise__

Install missing R packages.

```{r ini_install}
inst <- rownames(utils::installed.packages())
cran <- c("R.utils","refGenome","glmnet")
if(!all(cran %in% inst)){
    for(i in seq_along(cran)){
        if(!cran[i] %in% inst){
            install.packages(cran[i])
        }
    }
}

bioc <- c("snpStats","globaltest","rtracklayer")
if(!all(bioc %in% inst)){
    source("http://bioconductor.org/biocLite.R")
    for(i in seq_along(bioc)){
        if(!bioc[i] %in% inst){
            biocLite(bioc[i])
        }
    }
}
```

Verify PLINK installation.

```{r ini_plink}
setwd("/Users/gino/Documents/project3/data_plink")
if(system('plink -version') != 0){
    utils::browseURL("https://www.cog-genomics.org/plink2")
    stop("PLINK not found!")
}
```

Define the working directory.

```{r ini_path}
path <- "/Users/gino/Documents/project3/data_plink"
```

Choose the chromosome.

```{r ini_chr}
# chr <- 1
```

********************************************************************************

# __2. Download__

Download expression data:

```{r get_expression}
site <- "https://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1"
file <- "GD462.GeneQuantRPKM.50FN.samplename.resk10.txt.gz"
url <- file.path(site,file)
destfile <- file.path(path,file)
utils::download.file(url=url,destfile=destfile,method="auto")
R.utils::gunzip(filename=destfile,remove=FALSE,overwrite=TRUE)
```

Download sample annotation:

```{r get_samples}
site <- "http://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1"
file <- "E-GEUV-1.sdrf.txt"
url  <- file.path(site,file)
destfile <- file.path(path,file)
utils::download.file(url=url,destfile=destfile,method="libcurl")
```

```{r get_genes}
## Gene annotation: Unknown release of reference genome hg19!
# site <- paste0("ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/")
# file <- paste0("Homo_sapiens.GRCh37.75.gtf.gz")
# url  <- file.path(site,file)
# destfile <- file.path(path,file)
# utils::download.file(url=url,destfile=destfile,method="auto")
# R.utils::gunzip(filename=destfile,remove=FALSE,overwrite=TRUE)
###############################################################################
mart <- biomaRt::useMart(biomart="ensembl",dataset="hsapiens_gene_ensembl")
name <- substr(Y$map$TargetID,start=1,stop=15)
genesAnn <- biomaRt::getBM(attributes=c("ensembl_gene_id","hgnc_symbol","chromosome_name","start_position","end_position"), filters ="ensembl_gene_id",values =name,mart=mart)
martSnp <- biomaRt::useEnsembl(biomart="ENSEMBL_MART_SNP",dataset="hsapiens_snp")
name = head(X$map$snp.name)
###############################################################################

gtf_dir <- "gencode.v12.annotation.gtf"
gtf <- import(gtf_dir)
head(gtf)

### Keep protein coding genes only
index <- mcols(gtf)$gene_type == "protein_coding" & mcols(gtf)$type == "gene"
gtf2  <- gtf[index]
P38MAPK_path <- c("ATF1", "ATF2", "ATF6", "CEBPB", "CREB1", "CSNK2A1", "CSNK2A2", "CSNK2B", "DDIT3", "EIF4E", "EIF4EBP1", "ELK4", "ESR1", "GDI1", "HBP1", "HSPB1", "JUN", "KRT19", "KRT8", "MAPK11", "MAPK14", "MAPKAPK2", "MAPKAPK3", "MAPKAPK5", "MEF2A", "MEF2C", "MITF", "MKNK1", "NOS2", "PLA2G4A", "PPARGC1A", "PTGS2", "RAB5A", "RPS6KA4", "RPS6KA5", "SLC9A1", "TP53", "USF1", "BLK", "CCM2", "CDC42", "DUSP1", "DUSP10", "DUSP16", "DUSP8", "FGR", "FYN", "HCK", "LCK", "LYN", "MAP2K3", "MAP2K4", "MAP2K6", "MAP3K12", "MAP3K3", "PAK1", "PAK2", "PAK3", "PRKG1", "RAC1", "RALA", "RALB", "RIPK1", "SRC", "TAB1", "TRAF6", "YES1", "CCND1", "EEF2K", "MAPK12", "MAPK13", "MAPT", "PKN1", "SNTA1", "STMN1", "AC013461.1", "CDC25B", "ETV1", "LSP1", "RAF1", "SFN", "SRF", "TCF3", "TH", "TSC2", "YWHAB", "YWHAE", "YWHAG", "YWHAH", "YWHAQ", "YWHAZ", "ATM", "CAMK2B", "GADD45A", "GADD45B", "GADD45G", "MAP3K1", "MAP3K10", "MAP3K4", "MAP3K5", "MAP3K6", "MAP3K7", "TAB2", "TAOK1", "TAOK2", "TAOK3", "TRAF2", "TXN")

P38MAPK <- c("ATF1", "ATF2", "ATF6", "CEBPB", "CREB1", "CSNK2A1", "CSNK2A2", "CSNK2B", "DDIT3", "EIF4E", "EIF4EBP1", "ELK4", "ESR1", "GDI1", "HBP1", "HSPB1", "JUN", "KRT8", "MAPK11", "MAPK14", "MAPKAPK2", "MAPKAPK3", "MAPKAPK5", "MEF2A", "MEF2C", "MITF", "MKNK1", "NOS2", "PLA2G4A", "PPARGC1A", "PTGS2", "RAB5A", "RPS6KA4", "RPS6KA5", "SLC9A1", "TP53", "USF1", "BLK", "CCM2", "CDC42", "DUSP1", "DUSP10", "DUSP16", "DUSP8", "FGR", "FYN", "HCK", "LCK", "LYN", "MAP2K3", "MAP2K4", "MAP2K6", "MAP3K12", "MAP3K3", "PAK1", "PAK2", "PRKG1", "RAC1", "RALA", "RALB", "RIPK1", "SRC", "TAB1", "TRAF6", "YES1", "CCND1", "EEF2K", "MAPK12", "MAPK13", "MAPT", "PKN1", "SNTA1", "STMN1", "AC013461.1", "CDC25B", "LSP1", "RAF1", "SFN", "SRF", "TCF3", "TSC2", "YWHAB", "YWHAE", "YWHAG", "YWHAH", "YWHAQ", "YWHAZ", "ATM", "CAMK2B", "GADD45A", "GADD45B", "GADD45G", "MAP3K1", "MAP3K10", "MAP3K4", "MAP3K5", "MAP3K6", "MAP3K7", "TAB2", "TAOK1", "TAOK2", "TAOK3", "TRAF2", "TXN")

sum(mcols(gtf2)$gene_name %in% P38MAPK)
sum(P38MAPK %in% mcols(gtf2)$gene_name)
# P38MAPK_path[which(!P38MAPK_path %in% mcols(gtf2)$gene_name)]
# which(substr(mcols(gtf2)$gene_id, start=1,stop=15)=="ENSG00000091436")
# which(substr(mcols(gtf2)$gene_id, start=1,stop=15)=="ENSG00000224330")
# which(mcols(gtf2)$gene_name=="AC005019.2")
# which(mcols(gtf2)$gene_name=="AC013461.1")
# mcols(gtf2)[2863,]
# gtf2[2863,]
# AC013461.1 = ZAK
gtf_P38MAPK <- gtf2[mcols(gtf2)$gene_name %in% P38MAPK]
P38MAPK_id  <- substr(mcols(gtf_P38MAPK)$gene_id, start=1,stop=15)
head(mcols(gtf_P38MAPK))
head(mcols(gtf_P38MAPK@ranges))
df_iranges <- data.frame(iranges = gtf_P38MAPK@ranges)
sum(df_iranges$iranges.start %in% Y$map$Coord)
sum(Y$map$Coord %in% df_iranges$iranges.start)
#####
P38MAPK_id %in% substr(as.character(Y$data$TargetID), start=1, stop=15)
index2      <- match(substr(as.character(Y$data$TargetID), start=1, stop=15), P38MAPK_id)
P38MAPK_id  <- P38MAPK_id[index2]

df_gtf_P38MAPK <- mcols(gtf_P38MAPK)[index2,]
df_iranges  <- df_iranges[index2,]
```

Load and order expression data.

```{r pro_expression}
Y <- list()
Y$data <- utils::read.table(file=file.path(path,"GD462.GeneQuantRPKM.50FN.samplename.resk10.txt"),header=TRUE,nrows=-1)
Y$data <- Y$data[order(Y$data$Coord),]
Y$data <- Y$data[order(Y$data$Chr),]
Y$data <- Y$data[substr(as.character(Y$data$TargetID), start=1, stop=15) %in% P38MAPK_id,]
Y$map  <- Y$data[,c(1:4)]
Y$map_gtf  <- df_gtf_P38MAPK
Y$exData <- t(as.matrix(Y$data[,-c(1:4)]))
Y$map$Chr
substr(as.character(Y$map$TargetID), start=1, stop=15) == P38MAPK_id  
intersect(df_iranges$iranges.start, Y$map$Coord)
# path_geneId[which(!path_geneId %in% substr(as.character(Y$data$TargetID), start=1, stop=15))]
# gtf_pathAnno[which(!path_geneId %in% substr(as.character(Y$data$TargetID), start=1, stop=15))]

# which(substr(as.character(Y$data$TargetID), start=1, stop=15)=="ENSG00000203879")
ENSG00000203879
which(Y$map$Chr=="X")
# Y$map[23606,]
# sum(path_geneId %in% substr(as.character(Y$data$TargetID), start=1, stop=15))
# Y$data <- Y$data[Y$data$Chr==chr,] # test this???
for(i in 1:22){
  indices <- which(Y$map$Chr==i)
  if(length(indices)>=1){
    for(j in indices){
      chr <- i
      ## Download genotype data for chromosome i
      site <- "http://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/genotypes"
      file <- paste0("GEUVADIS.chr",chr,".PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz")
      url <- file.path(site,file)
      destfile <- file.path(path,file)
      utils::download.file(url=url,destfile=destfile,method="libcurl")
      
      ## Exclude SNPs with $\text{maf}<5\%$ or $\text{geno}>0\%$
      setwd(path)
      system(paste0("plink --vcf GEUVADIS.chr",chr,".PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz        --maf 0.05 --geno 0 --make-bed --out snps",chr),invisible=FALSE)
      
      chr <- 22
      ## Load genotype data
      library(snpStats)
      bed <- file.path(path,paste0("snps",chr,".bed"))
      bim <- file.path(path,paste0("snps",chr,".bim"))
      fam <- file.path(path,paste0("snps",chr,".fam"))
      X   <- snpStats::read.plink(bed=bed,bim=bim,fam=fam)
      X$fam <- NULL
      names(X)[1] <- "data"
      all(diff(X$map$position) > 0)
      
      ##Exclude samples without expression or genotype data
      intersect  <- intersect(rownames(Y$eData),rownames(X$data))
      Y$eData   <- Y$eData[intersect,]
      X$data     <- X$data[intersect,]
      
      ## Exclude samples from population YRI
      samples <- utils::read.delim(file=file.path(path,"E-GEUV-1.sdrf.txt"))
      match   <- match(rownames(Y$eData),samples$Source.Name)
      cond    <- samples$Factor.Value.population[match]!="YRI"
      Y$eData  <- Y$eData[cond,]
      X$data  <- X$data[cond,]
      
      ## Exclude SNPs with $\text{maf}<5\%$
      maf <- snpStats::col.summary(X$data)$MAF
      cond <- maf >= 0.05
      X$data <- X$data[,cond]
      X$map <- X$map[cond,]
      
      ## Exclude genes with zero-variance
      cond <- apply(Y$eData,2,function(x) sd(x)>0)
      Y$eData <- Y$eData[,cond]
      Y$map <- Y$map[cond,]
    }
  } 
}    
```


Download genotype data:

```{r get_genotype}
site <- "http://www.ebi.ac.uk/arrayexpress/files/E-GEUV-1/genotypes"
file <- paste0("GEUVADIS.chr",chr,".PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz")
url <- file.path(site,file)
destfile <- file.path(path,file)
utils::download.file(url=url,destfile=destfile,method="libcurl")
```

Exclude SNPs with $\text{maf}<5\%$ or $\text{geno}>0\%$.

```{r get_selection}
setwd(path)
system(paste0("plink --vcf GEUVADIS.chr",chr,".PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz --maf 0.05 --geno 0 --make-bed --out snps",chr),invisible=FALSE)
```

********************************************************************************

## __3. Process__

Load genotype data:

```{r pro_genotype}
library(snpStats)
bed <- file.path(path,paste0("snps",chr,".bed"))
bim <- file.path(path,paste0("snps",chr,".bim"))
fam <- file.path(path,paste0("snps",chr,".fam"))
X   <- snpStats::read.plink(bed=bed,bim=bim,fam=fam)
X$fam <- NULL
names(X)[1] <- "data"
all(diff(X$map$position) > 0)
```

Exclude samples without expression or genotype data.

```{r pro_samples}
intersect <- intersect(rownames(Y$data),rownames(X$data))
Y$data    <- Y$data[intersect,]
X$data    <- X$data[intersect,]
```

Exclude samples from population YRI.

```{r pro_population}
samples <- utils::read.delim(file=file.path(path,"E-GEUV-1.sdrf.txt"))
match   <- match(rownames(Y$data),samples$Source.Name)
cond    <- samples$Factor.Value.population[match]!="YRI"
Y$data  <- Y$data[cond,]
X$data  <- X$data[cond,]
```

Exclude SNPs with $\text{maf}<5\%$.

```{r pro_frequency}
maf <- snpStats::col.summary(X$data)$MAF
cond <- maf >= 0.05
X$data <- X$data[,cond]
X$map <- X$map[cond,]
```

Exclude genes with zero-variance.

```{r pro_variance}
cond <- apply(Y$data,2,function(x) sd(x)>0)
Y$data <- Y$data[,cond]
Y$map <- Y$map[cond,]
```

********************************************************************************

## __4. Match__

```{r mat_annot}
## Identify all protein-coding genes:
# map <- list()
# setwd(path)
# object <- refGenome::ensemblGenome()
# refGenome::basedir(object) <- file.path(path)
# filename <- paste0("Homo_sapiens.GRCh37.75.gtf")
# refGenome::read.gtf(object,filename=filename)
# map <- refGenome::getGenePositions(object=object,by="gene_id")
# map <- map[map$seqid==chr & map$gene_biotype=="protein_coding",
#            c("gene_id","seqid","start","end")]
## Unknown release of reference genome hg19!
```

Identify SNPs within one million base pairs around gene start.

```{r mat_compress}
p <- ncol(Y$data)
match <- data.frame(from=integer(length=p),to=integer(length=p))

pb <- utils::txtProgressBar(min=0,max=p,style=3)
for(i in seq_len(p)){ # decrease p to sum(Y$map$Chr==1)
    utils::setTxtProgressBar(pb=pb,value=i)
    
    # conditions: same chromosome, within window
    cond1 <- X$map$chromosome == Y$map$Chr[i]
    if(!any(cond1)){
        stop("Wrong chromosome!")
    }
    cond2 <- X$map$position >= Y$map$Coord[i] - 1*10^6 # change!
    cond3 <- X$map$position <= Y$map$Coord[i] + 1*10^6 # change!
    which <- as.integer(which(cond1 & cond2 & cond3))
    
    # compression: first and last entry
    if(length(which)==0){next}
    match$from[i] <- min(which)
    match$to[i] <- max(which)
    if(length(which)==1){next}
    if(!all(diff(which)==1)){
        stop("Wrong order!")
    }
}
```

********************************************************************************

## __5. Predict__

Regress gene expression on SNPs (example).

```{r pre_lasso}
i <- sample(which(Y$map$Chr==chr),size=1) # index
if(Y$map$Chr[i]!=chr){
    stop("Wrong chromosome!")
}
y <- Y$data[,i]
if((match$to[i]-match$from[i]<2)){
    warning("Too few covariates!")
} 
x <- X$data[,seq(from=match$from[i],to=match$to[i],by=1)]
x <- abs(as(x,"numeric")-2) # major/minor allele
net <- glmnet::cv.glmnet(y=y,x=x)
```
