% load packages
<<packages, include=FALSE, echo=FALSE>>=
library(cambridge)
@

% set options
<<settings, include=FALSE, echo=FALSE>>=
opts_knit$set(root.dir="..", base.dir="../figs")
opts_chunk$set(fig.align='center', fig.path="../figs/", 
               echo=FALSE, cache=FALSE, message=FALSE, fig.pos='!ht')
@

% read figure code
<<figures, include=FALSE>>=
read_chunk("code/figures.R")
@

\documentclass[a4paper,hidelinks]{article}
\usepackage{url,bm,algorithm,algpseudocode,bbm,amsmath,amssymb,amsthm,
amsfonts,graphics,graphicx,epsfig,rotating,caption,subcaption,natbib,
titlesec,multicol,verbatim,threeparttable,booktabs,dsfont,parskip,
xr-hyper,hyperref,enumerate}
\externaldocument[sm-]{supplement}

% vectors and matrices
\newcommand{\bbeta}{\bm{\beta}}
\newcommand{\btau}{\bm{\tau}}
\newcommand{\bsigma}{\bm{\sigma}}
\newcommand{\balpha}{\bm{\alpha}}
\newcommand{\bepsilon}{\bm{\epsilon}}
\newcommand{\bomega}{\bm{\omega}}
\newcommand{\bOmega}{\bm{\Omega}}
\newcommand{\bSigma}{\bm{\Sigma}}
\newcommand{\bkappa}{\bm{\kappa}}
\newcommand{\btheta}{\bm{\theta}}
\newcommand{\bmu}{\bm{\mu}}
\newcommand{\bpsi}{\bm{\psi}}
\newcommand{\blambda}{\bm{\lambda}}
\newcommand{\bLambda}{\bm{\Lambda}}
\newcommand{\bPsi}{\bm{\Psi}}
\newcommand{\bphi}{\bm{\phi}}
\newcommand{\Y}{\mathbf{Y}}
\newcommand{\y}{\mathbf{y}}
\newcommand{\X}{\mathbf{X}}
\newcommand{\x}{\mathbf{x}}
\newcommand{\I}{\mathbf{I}}
\newcommand{\bgamma}{\bm{\gamma}}
\newcommand{\T}{\mathbf{T}}
\newcommand{\Z}{\mathbf{Z}}
\newcommand{\W}{\mathbf{W}}
\renewcommand{\O}{\mathbf{O}}
\newcommand{\B}{\mathbf{B}}
\renewcommand{\H}{\mathbf{H}}
\newcommand{\U}{\mathbf{U}}
\newcommand{\Em}{\mathbf{E}}
\newcommand{\D}{\mathbf{D}}
\newcommand{\Vm}{\mathbf{V}}
\newcommand{\rv}{\mathbf{r}}
\newcommand{\A}{\mathbf{A}}
\newcommand{\Q}{\mathbf{Q}}
\newcommand{\Sm}{\mathbf{S}}
\newcommand{\0}{\bm{0}}
\newcommand{\tx}{\tilde{\mathbf{x}}}
\newcommand{\hy}{\hat{y}}
\newcommand{\tm}{\tilde{m}}
\newcommand{\tkappa}{\tilde{\kappa}}
\newcommand{\m}{\mathbf{m}}
\newcommand{\C}{\mathbf{C}}
\renewcommand{\v}{\mathbf{v}}
\newcommand{\g}{\mathbf{g}}
\newcommand{\s}{\mathbf{s}}
\renewcommand{\c}{\mathbf{c}}
\newcommand{\ones}{\mathbf{1}}
\newcommand{\R}{\mathbf{R}}
\newcommand{\rvec}{\mathbf{r}}
\newcommand{\Lm}{\mathbf{L}}

% functions and operators
\renewcommand{\L}{\mathcal{L}}
\newcommand{\G}{\mathcal{G}}
\renewcommand{\P}{\mathcal{P}}
\newcommand{\argmax}{\text{argmax} \,}
\newcommand{\expit}{\text{expit}}
\newcommand{\erfc}{\text{erfc}}
\newcommand{\E}{\mathbb{E}}
\newcommand{\V}{\mathbb{V}}
\newcommand{\Cov}{\mathbb{C}\text{ov}}
\newcommand{\tr}{^{\text{T}}}
\newcommand{\diag}{\text{diag}}
\newcommand{\KL}{D_{\text{KL}}}
\newcommand{\trace}{\text{tr}}
\newcommand{\norm}[1]{\left\lVert #1 \right\rVert}

% distributions
\newcommand{\N}{\text{N}}
\newcommand{\TG}{\text{TG}}
\newcommand{\Bi}{\text{Binom}}
\newcommand{\PG}{\text{PG}}
\newcommand{\Mu}{\text{Multi}}
\newcommand{\GIG}{\text{GIG}}
\newcommand{\IGauss}{\text{IGauss}}
\newcommand{\Un}{\text{U}}

% syntax and readability
\renewcommand{\(}{\left(}
\renewcommand{\)}{\right)}
\renewcommand{\[}{\left[}
\renewcommand{\]}{\right]}

% settings
\graphicspath{{/Users/magnusmunch/Documents/OneDrive/PhD/project_cambridge/graphs/}}
\setlength{\evensidemargin}{.5cm} \setlength{\oddsidemargin}{.5cm}
\setlength{\textwidth}{15cm}
\title{Drug efficacy prediction in cell lines}
\date{\today}
\author{Magnus M. M\"unch$^{1,2}$\footnote{Correspondence to: \href{mailto:m.munch@vumc.nl}{m.munch@vumc.nl}}, Mark A. van de Wiel$^{1,3}$, Sylvia Richardson$^{3}$, \\ and Gwena{\"e}l G. R. Leday$^{3}$}

\begin{document}

	\maketitle
	
	\noindent
	1. Department of Epidemiology \& Biostatistics, Amsterdam Public Health research institute, VU University Medical Center, PO Box 7057, 1007 MB
	Amsterdam, The Netherlands\\
	2. Mathematical Institute, Leiden University, Leiden, The Netherlands \\
	3. MRC Biostatistics Unit, Cambridge Institute of Public Health, Cambridge, United Kingdom \\
	
	\begin{abstract}
		{...}
	\end{abstract}
	
	\noindent\textbf{Keywords}: 
	
	\noindent\textbf{Software available from}:
	
	\section{Introduction}
	- motivation: drug response prediction from different omics data types is 
	promising both for precision medicine and drug target discovery and relates to
	actual tumor response \cite[]{iorio_landscape_2016}. 
		
	- problem: prediction of drug efficacy in cell lines is difficult (DREAM 7 
	challenge in \cite{costello_community_2014}).	

	- example data: introduction (GDSC, CCLE, NCI-60)
	
	- feature information: prior experts, pathways, database summaries, previous 
	experiments, type of feature (somatic mutations, copy number alterations, DNA 
	methylation, and gene expression). 
	
	- drugs information: molecular target, drug class, pubchem, 
	omnipath DB, pharmaco DB
	
	- drugs x feature information
	
	- current literature: reviews in \cite{azuaje_computational_2017} and
	\cite{ali_machine_2019}, 
	full multivariate approach by Oslo people in \cite{mai_composite_2019}, 
	TANDEM \cite[]{aben_tandem:_2016}, DeltaNet 
	by \cite{noh_inferring_2016}, patient-derived tumor cell lines 
	in \cite{gao_high-throughput_2015}, 
	multiple kernel learning approach in \cite{costello_community_2014},
	\cite{ammad-ud-din_drug_2016} combine a 
	multiple kernel learning approach with matrix factorization, using the model 
	to uncover the latent 
	relationships between drug targets and intracellular pathways, and
	structured penalized regression in \cite{zhao_structured_2019}

	
	- our solution: adaptive borrowing of information through empirical Bayes, 
	guided by external feature and drug information.
	
	\section{Methods}
	\subsection{Simultaneous equations model}\label{sec:SEM}
	Assume we have continuous efficacy measures $y_{id}$ for cell lines 
	$i=1,\dots, n$, centred per drug $d$. We predict efficacy from molecular 
	features 
	$x_{ij}$, $j=1,\dots, p$, collected in $\x_i = \begin{bmatrix} x_{i1} & 
	\dots & x_{ip} \end{bmatrix} \tr$. We have an \textit{a priori} (possibly 
	informative) partitioning 
	of the omics features into $g=1, \dots, G$ non-overlapping groups.
	For convenience, we let 
	$g(j)$ denote the group index of molecular feature $j$. Furthermore, we have 
	characteristics of the drugs available in the form of external variables 
	$\mathbf{C} = \begin{bmatrix} \mathbf{c}_1 & \dots & \mathbf{c}_D 
	\end{bmatrix} \tr$. 
	
	We regress the drug efficacies on the molecular feature effects $\beta_{jd}$:
	\begin{subequations}
		\begin{align}
		  y_{id} & = \x_i \tr \bbeta_d + \epsilon_{id}, \label{eq:linearmodel}\\
		  \text{with } & \epsilon_{id} \sim \mathcal{N}(0, \sigma_d^2),
		\end{align}
	\end{subequations}
	where $p$-dimensional $\bbeta_d$ are the drug-specific omics feature effects. 
	Note that (\ref{eq:linearmodel}) gives rise to a system of $d=1, \dots, D$ 
	linear simultaneous equations.
	
	\subsection{Bayesian prior model}
	We capture the uncertainty in the parameters through a Bayesian prior model
	and endow the parameters with the following prior:
	\begin{subequations}\label{eq:prior}
		\begin{align}
		  \beta_{jd} & \sim \mathcal{N}_p (0, \phi^2_{g(j)} \gamma_{jd}^2 
		  \sigma_d^2), 
		  \label{eq:betaprior}\\
		  \gamma_{jd}^{2} & \sim \mathcal{IG}( 
		  [(\mathbf{c}_d^j) \tr \bm{\alpha}]^{-1}, \lambda_d), \\
		  \sigma_d^{2} & \sim 1/\sigma_d^{3},
		\end{align}
	\end{subequations}
	where $\mathcal{IG}(\theta, \lambda)$ denotes an inverse Gaussian distribution 
	with mean $\theta$ and shape $\lambda$. The different molecular feature types
	are given their own weights through the $\phi^2_{g}$, while the external 
	outcome-feature data comes in through 
	our mean model for the $\gamma_{jd}^2$ hyperprior model: 
	$[(\mathbf{c}_d^j) \tr \bm{\alpha}]^{-1}$. Both the $\phi^2_{g}$ and
	$\bm{\alpha}$ are estimated in a data-driven manner. More details on the model
	itself, as well as its estimation, may be found in 
	the SM and in [! ref to first paper here].
	
	\section{Data}
	
	\subsection{GDSC}
	The GDSC project's \cite[]{yang_genomics_2013} aim is "to improve cancer 
	treatments by discovering therapeutic biomarkers that can be used to identify 
	patients most likely to respond to anticancer drug". Part of the project is to
	screen $>1000$ human cancer cell lines with drugs. The cell lines have been
	genetically characterised and several drug sensitivity measures are recorded.
	The data is freely available from \cite{garnett_systematic_2012} and consist 
	of: (i) the sensitivity measures 
	(mainly half maximal inhibitory concentration, i.e., IC50) of the cell lines 
	to the drugs, (ii) annotation of the 
	screened compounds, and (iii) the cell lines' genomic profile (mutations, 
	copy numbers, methylation profiles, and expression). We will attempt to 
	predict drug sensitivity of the cell lines, as quantified by IC50, using the
	genomic data. Other choices of efficacy measures than IC50 are possible, 
	but a discussion on the pros and cons of these different choices is not the 
	aim of this paper. 
	
	\subsubsection{Primary data}
	We took the logarithm of the IC50 values and took the average if there were
	multiple measurements per cell line-drug combination. In the following, if we 
	refer to IC50, it denotes the log-transformed values. After iteratively 
	removing the cell line or drug with the highest proportion 
	of missings until there are no more missing values, we end up with IC50 
	estimates for 482 cell lines on 168 drugs. For the gene expression 
	data, we fit a lasso with IC50 as response to the drugs separately and 
	adjusted the penalty parameter to select approximately 300 genes per drug.
	
	- From Oslo:
	1.2) Preselecting GEX features: we selected gene expression features
	explaining 50\% variations over all cell lines, since this results in not too
	features and probably remains those important features.
	1.3) Preselecting CNV: we used the same strategy as Garnett et al. (2012)
	where the GDSC data are from.
	1.4) Preselecting MUT: we deleted mutated genes with less than two cell lines.
	
	- check out selection method in \cite{zhao_structured_2019}.
	
	- how to transform the data and deal with missingness? Read GDSC papers.
	
	\subsubsection{External data}
	Several external drug variables are available: (i) the developmental stage
	(experimental, in clinical development, or clinically approved) and (ii)
	the action (cytotoxic or targeted). We also have a external variable 
	available that relates the drugs to the features: a binary indicator whether 
	a gene is in the target pathway of the drug or not.
  
  - check whether compounds are PAINS \cite[]{baell_new_2010} or not. 
  PAINS indicates whether compound is frequent false hitter.
  
  - clinical trials database CTRP might contain info

  - extra data on GDSC might be in PharamcoGx R-package. Also contains bunch 
  of functions.
  
  - check GSK open target EBI or ask Chris Wallace about it. Database that links
	targets to drugs and diseases. Might be useful for drugs co-data.
	
	- try non-negative matrix factorisation/independent component analysis on 
	binary fingerprints to get principal components that characterise the 
	different drugs.  
	
	\subsection{CCLE}
	- extra data on CCLE might be in PharamcoGx R-package
	
	\subsection{NCI-50}
	
	\subsection{CMap}
	- Cmap \cite[]{subramanian_next_2017,lamb_connectivity_2006} contains 
	pharma-genomic data (similar to CCLE and GDSC). 
	
	- Data available at: 
	https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742. 
	
	- Cmap website: https://clue.io/cmap.
	
	\section{Results}
	\subsection{GDSC}
	
	
	\section{Discussion}
	- use copula's and marginal posteriors to find multivariate posteriors for 
	contrast testing.
	
	- empirical Bayes gives extra information via estimation of the 
	hyperparameters, such as informativeness of the drug information (via model 
	for the $\gamma_d$) and relatedness of the tissues (via $\bm{\Omega}$ and 
	$\nu$)
	
	- regression on drug covariates very intuitive.

	\section*{Supplementary Material}
	Supplementary Material is available online from 
	\url{https://github.com/magnusmunch/cambridge}. 
	
	\section*{Reproducible Research}
	All results and documents may be recreated from 
	\url{https://github.com/magnusmunch/cambridge}.
	
	\section*{Acknowledgements}
	We thank Zhi Zhao and Leiv Tore Salte R{\o}nneberg from the Oslo Centre for 
	Biostatistics and Epidemiology for help with the pre-processing of the data. 
	\textit{Conflict of Interest}: None declared.
	
	\bibliographystyle{author_short3} 
	\bibliography{refs}
	
	\section*{Session info}

<<r session-info, eval=TRUE, echo=TRUE, tidy=TRUE>>=
devtools::session_info()
@

\end{document}